File: leetcode/problems/language-specific/go/49.group-anagrams.go
/*
 * @lc app=leetcode id=49 lang=golang
 *
 * [49] Group Anagrams
 */

// @lc code=start
func groupAnagrams(strs []string) [][]string {
	if len(strs) <= 0 {
		return [][]string{}
	}

	res, mappings := [][]string{}, map[string][]string{}
	for _, v := range strs {
		characterMappings := make([]int, 26)
		for _, v1 := range v {
			characterMappings[v1-'a']++
		}

		keyString := ""
		for _, v1 := range characterMappings {
			keyString += string(v1)
		}

		mappings[keyString] = append(mappings[keyString], v)
	}

	for _, v := range mappings {
		res = append(res, v)
	}

	return res
}

// @lc code=end

File: leetcode/problems/sessions/2025-10-25_2025-??-??/49.group-anagrams.py
#
# @lc app=leetcode id=49 lang=python3
#
# [49] Group Anagrams
#

# @lc code=start

# * Hash Table Solution | O(mn) Time | O(n) Space
# * m -> The length of strs array | n -> The length of each string


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = defaultdict(list)
        for word in strs:
            word_counter = [0] * 26
            for char in word:
                word_counter[ord(char) - ord("a")] += 1
            res[tuple(word_counter)].append(word)

        return list(res.values())


# @lc code=end
